### PR TITLE
add resource dms_kafka_instance

### PR DESCRIPTION
**Pull request by Jason-Zhang9309**
**What this PR does / why we need it**:

add resource dms_kafka_instance

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
related #1052 

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud' TESTARGS='-run TestAccDmsKafkaInstances'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run TestAccDmsKafkaInstances -timeout 360m -parallel 4
=== RUN   TestAccDmsKafkaInstances_basic
=== PAUSE TestAccDmsKafkaInstances_basic
=== RUN   TestAccDmsKafkaInstances_withEpsId
=== PAUSE TestAccDmsKafkaInstances_withEpsId
=== CONT  TestAccDmsKafkaInstances_basic
=== CONT  TestAccDmsKafkaInstances_withEpsId
--- PASS: TestAccDmsKafkaInstances_withEpsId (463.21s)
--- PASS: TestAccDmsKafkaInstances_basic (484.29s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       484.337s
```

### CODE REVIEW

File: docs/resources/dms_kafka_instance.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    consists of 4 to 64 characters, and supports only letters, digits, hyphens (-) and underscores (_).
+
+* `description` - (Optional, String) Specifies the description of the DMS kafka instance. It is a character
+    string containing not more than 1024 characters.
+
+* `engine_version` - (Required, String, ForceNew) Specifies the version of the kafka engine.
+    Valid values are "1.1.0" and "2.3.0".
+    Changing this creates a new instance resource.
+
+* `specification` - (Required, String, ForceNew) Specifies the baseline bandwidth of the DMS kafka instance,
```
Comment (ShiChangkuo) [inline docs/resources/dms_kafka_instance.md]: can we rename it to `bandwidth` to keep with the console?

File: docs/resources/dms_kafka_instance.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    Changing this creates a new instance resource.
+
+* `security_group_id` - (Required, String) Specifies the ID of a security group.
+
+* `available_zones` - (Required, List, ForceNew) Specifies the ID of an AZ. The parameter value can not be
+    left blank or an empty array. Changing this creates a new instance resource.
+
+* `product_id` - (Required, String, ForceNew) Specifies a product ID. Changing this creates a new instance resource.
+
+* `kafka_manager_user` - (Required, String, ForceNew) Specifies the username for logging in to the Kafka Manager.
```
Comment (ShiChangkuo) [inline docs/resources/dms_kafka_instance.md]: suggest to rename it to `manager_user`

File: docs/resources/dms_kafka_instance.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    - When specification is 1200MB: 8
+
+    Changing this creates a new instance resource.
+
+* `retention_policy` - (Optional, String) Specifies the action to be taken when the memory usage reaches
+    the disk capacity threshold. Value range:
+    - `time_base`: Automatically delete the earliest messages.
+    - `produce_reject`: Stop producing new messages.
+
+* `connector_enable` - (Optional, Bool, ForceNew) Specifies whether to enable dumping.
```
Comment (ShiChangkuo) [inline docs/resources/dms_kafka_instance.md]: suggest rename it to `dumping` to keep with the console

File: docs/resources/dms_kafka_instance.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    Changing this creates a new instance resource.
+
+* `enable_auto_topic` - (Optional, Bool, ForceNew) Specifies whether to enable automatic topic creation. Value range:
+    - `true`: enable
+    - `false`: disable
```
Comment (ShiChangkuo) [inline docs/resources/dms_kafka_instance.md]: we can set the value to `true` of `false` for a bool type, so the two lines are redundant.

File: docs/resources/dms_kafka_instance.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+* `storage_spec_code` - (Required, String, ForceNew) Specifies the storage I/O specification. Value range:
+    - When specification is 100MB: dms.physical.storage.high or dms.physical.storage.ultra
+    - When specification is 300MB: dms.physical.storage.high or dms.physical.storage.ultra
+    - When specification is 600MB: dms.physical.storage.ultra
+    - When specification is 1200MB: dms.physical.storage.ultra
+
+    Changing this creates a new instance resource.
+
+* `partition_num` - (Required, Int, ForceNew) Specifies the maximum number of topics in the DMS kafka instance.
```
Comment (ShiChangkuo) [inline docs/resources/dms_kafka_instance.md]: if the value depends on `specification`, we can use it as an internal parameter or set it as an attribute

File: huaweicloud/resource_huaweicloud_dms_kafka_instance.go [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+			"vpc_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"security_group_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"subnet_id": {
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_dms_kafka_instance.go]: use `network_id` instead

File: huaweicloud/resource_huaweicloud_dms_kafka_instance.go [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+			},
+			"kafka_manager_user": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"kafka_manager_password": {
+				Type:      schema.TypeString,
+				Required:  true,
+				Sensitive: true,
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_dms_kafka_instance.go:101]: missing `RequiredWith`

File: huaweicloud/resource_huaweicloud_dms_kafka_instance.go [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+				Computed: true,
+				ForceNew: true,
+			},
+			"enable_auto_topic": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"storage_spec_code": {
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_dms_kafka_instance.go]: put the **required** field in the head of the schema

File: docs/resources/dms_kafka_instance.md [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+* `used_storage_space` - Indicates the used message storage space. Unit: GB
+* `connect_address` - Indicates the IP address of the DMS kafka instance.
+* `port` - Indicates the port number of the DMS kafka instance.
+* `status` - Indicates the status of the DMS kafka instance.
+* `ssl_enable` - Indicates whether the SSL-encrypted access is enabled.
+* `enable_publicip` - Indicates whether public access to the DMS kafka instance is enabled.
+* `resource_spec_code` - Indicates a resource specifications identifier.
+* `type` - Indicates the DMS kafka instance type.
+* `user_id` - Indicates a user ID.
+* `user_name` -	Indicates a username.
```
Comment (ShiChangkuo) [inline docs/resources/dms_kafka_instance.md]: tenant user id or the kafka user id?

File: huaweicloud/resource_huaweicloud_dms_kafka_instance.go [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+			},
+			"user_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"user_name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"manegement_connect_address": {
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_dms_kafka_instance.go:207]: `connect_address` please

File: huaweicloud/resource_huaweicloud_dms_kafka_instance.go [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+			},
+		},
+	}
+}
+
+func resourceDmsKafkaPublicIpIDs(d *schema.ResourceData) (string, error) {
+	specification := d.Get("specification").(string)
+	publicIpIDsRaw := d.Get("public_ip_ids").([]interface{})
+
+	if specification == "100MB" && len(publicIpIDsRaw) != 3 {
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_dms_kafka_instance.go]: using a map will be more clearer

File: docs/resources/dms_kafka_instance.md (hunk 2 above)
Comment (Jason-Zhang9309) [reply docs/resources/dms_kafka_instance.md]: fixed

File: docs/resources/dms_kafka_instance.md (hunk 3 above)
Comment (Jason-Zhang9309) [reply docs/resources/dms_kafka_instance.md]: fixed

File: docs/resources/dms_kafka_instance.md (hunk 4 above)
Comment (Jason-Zhang9309) [reply docs/resources/dms_kafka_instance.md]: fixed

File: huaweicloud/resource_huaweicloud_dms_kafka_instance.go (hunk 6 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_dms_kafka_instance.go]: fixed

File: huaweicloud/resource_huaweicloud_dms_kafka_instance.go (hunk 7 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_dms_kafka_instance.go:101]: `manager_user` and `manager_password` are both required params, no need for `RequiredWith`

File: huaweicloud/resource_huaweicloud_dms_kafka_instance.go (hunk 8 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_dms_kafka_instance.go]: fixed

File: huaweicloud/resource_huaweicloud_dms_kafka_instance.go (hunk 10 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_dms_kafka_instance.go:207]: There are too different attributes.
`connect_address`: Indicates the IP address of an instance. For example, "connect_address": "192.168.1.116,192.168.1.152,192.168.1.78"
`management_connect_address`: Indicates the connection address of the Kafka Manager of a Kafka instance. For example, "management_connect_address": "https://192.168.1.116:9999"

File: docs/resources/dms_kafka_instance.md (hunk 9 above)
Comment (Jason-Zhang9309) [reply docs/resources/dms_kafka_instance.md]: fixed

File: docs/resources/dms_kafka_instance.md (hunk 5 above)
Comment (Jason-Zhang9309) [reply docs/resources/dms_kafka_instance.md]: fixed

File: huaweicloud/resource_huaweicloud_dms_kafka_instance.go (hunk 11 above)
Comment (Jason-Zhang9309) [reply huaweicloud/resource_huaweicloud_dms_kafka_instance.go]: fixed

File: docs/resources/dms_kafka_instance.md (hunk 1 above)
Comment (Jason-Zhang9309) [reply docs/resources/dms_kafka_instance.md]: fixed

**Comment by Jason-Zhang9309** [comment]
latest test result:
```
make testacc TEST='./huaweicloud' TESTARGS='-run TestAccDmsKafkaInstances'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run TestAccDmsKafkaInstances -timeout 360m -parallel 4
=== RUN   TestAccDmsKafkaInstances_basic
=== PAUSE TestAccDmsKafkaInstances_basic
=== RUN   TestAccDmsKafkaInstances_withEpsId
=== PAUSE TestAccDmsKafkaInstances_withEpsId
=== CONT  TestAccDmsKafkaInstances_basic
=== CONT  TestAccDmsKafkaInstances_withEpsId
--- PASS: TestAccDmsKafkaInstances_withEpsId (749.50s)
--- PASS: TestAccDmsKafkaInstances_basic (770.96s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       771.017s
```

**Pull request merged**